File: algorithms/stronglyConnectedComponents.cpp
#include <bits/stdc++.h>
using namespace std;

vector<bool> visited;

struct G {
    vector<vector<int>> g, reversed_g;
    vector<int> visited, top_sort_order;

    void input(){
        int n, m; cin >> n >> m;
        g.resize(n); reversed_g.resize(n);
        for (int i = 0; i < m; i++){
            int from, to; cin >> from >> to;
            from--; to--;
            g[from].push_back(to);
            reversed_g[to].push_back(from);
        }        
    }

    // Это dfs для топологической сортировки
    void dfs1(int v){
        visited[v] = 1;
        for (auto to : reversed_g[v]) if (!visited[to]) dfs1(to);

        top_sort_order.push_back(v);
    }
    
    // Это dfs для раскраски вершин
    void dfs2(int v, int component){
        visited[v] = component;
        for (auto to : g[v]) if (!visited[to]) dfs2(to, component);
    }

    vector<int> findSCC() {
        // Сперва ищем вершины, от которых будем запускать dfs для выделения компонент сильной связности
        visited.assign(g.size(), 0);
        for (int v = 0; v < g.size(); v++) if (!visited[v]) dfs1(v);

        reverse(top_sort_order.begin(), top_sort_order.end());

        // Затем красим вершины
        visited.assign(g.size(), 0);
        int component_number = 0;
        for(auto v: top_sort_order) if (!visited[v]){
            component_number++;
            dfs2(v, component_number);
        }
        
        // Для каждой вершины будет определен номер компоненты сильной связности, которой она принадлежит
        return visited;
    }
};


int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    G g = G();
    g.input();
    for(auto el : g.findSCC()){
        cout << el << ' ';
    }
}

File: algorithms/stronglyConnectedComponents.py
from sys import setrecursionlimit

n, m = map(int, input().split())
setrecursionlimit(m)

# Это dfs для топологической сортировки
def dfs1(v):
    visited[v] = 1
    for to in reversed_g[v]:
        if not visited[to]:
            dfs1(to)

    top_sort_order.append(v)

# Это dfs для раскраски вершин
def dfs2(v, component):
    visited[v] = component
    for to in g[v]:
        if not visited[to]:
            dfs2(to, component)
            

g = [[] for _ in range(n)]
reversed_g = [[] for _ in range(n)]
for _ in range(m):
    f, t = map(int, input().split())
    f -= 1
    t -= 1

    g[f].append(t)
    reversed_g[t].append(f)

# Сперва ищем вершины, от которых будем запускать dfs для выделения компонент сильной связности
visited = [0 for _ in range(n)]
top_sort_order = []
for v in range(n):
    if not visited[v]:
        dfs1(v)

top_sort_order.reverse()

# Затем красим вершины
visited = [0 for _ in range(n)]
component_number = 0
for v in top_sort_order:
    if not visited[v]:
        component_number += 1
        dfs2(v, component_number)

for v in visited:
    print(v, end=' ')
